server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -218,14 +218,85 @@ public void writeInt(int i) throws IOException {
      * using {@link #writeInt}
      */
     public void writeVInt(int i) throws IOException {
-        final byte[] buffer = scratch.get();
-        int index = 0;
-        while ((i & ~0x7F) != 0) {
-            buffer[index++] = ((byte) ((i & 0x7f) | 0x80));
-            i >>>= 7;
+        /*
+         * Pick the number of bytes that we need based on the value and then
+         * encode the int, unrolling the loops by hand. This allows writing
+         * small numbers to use `writeByte` which is simple and fast. The
+         * unrolling saves a few comparisons and bitwise operations. All
+         * together this saves quite a bit of time compared to a naive
+         * implementation.
+         */
+        switch (Integer.numberOfLeadingZeros(i)) {
```

Comment:
This gets compiled to [`lzcnt`](https://www.felixcloutier.com/x86/lzcnt) and the JVM's tableswitch. At this point the overhead of the `buffer` and `BigArrays` dominates the method.